File: test/fixtures/safe.rb
# Idiomatic, safe Ruby. Any finding here is a false positive.
require 'digest'

ALLOWED_HOSTS = ['api.internal.example'].freeze

def get_user(db, params)
  # Bound parameter — the driver escapes it.
  db.execute('SELECT * FROM users WHERE id = ?', [params[:id]])
end

def ping(params)
  # Argument list, so no shell parses the value.
  host = params[:host]
  system('ping', '-c', '1', host) if ALLOWED_HOSTS.include?(host)
end

def digest(data)
  Digest::SHA256.hexdigest(data)
end

AWS_ACCESS_KEY = ENV['AWS_ACCESS_KEY']

File: test/fixtures/safe.py
# Idiomatic, safe Python. Any finding here is a false positive.
import hashlib
import os
import subprocess
from flask import request

ALLOWED_HOSTS = {"api.internal.example"}


def get_user(cursor):
    # Parameterized — the driver binds the value.
    cursor.execute("SELECT * FROM users WHERE id = %s", (request.args.get('id'),))


def ping():
    # Argument list, no shell.
    host = request.args.get('host')
    if host in ALLOWED_HOSTS:
        subprocess.run(["ping", "-c", "1", host], shell=False, check=False)


def digest(data):
    return hashlib.sha256(data).hexdigest()


# This comment mentions a password = "example" but declares nothing.
API_KEY = os.environ.get("API_KEY")
